File: comgongdaeng/DynamicProgramming/BOJ_1106.py
import sys
import math

c, n = map(int, sys.stdin.readline().split())
# dp[i] i명의 고객을 늘리기 위한 최소 비용
dp = [math.inf] * (c+100)
dp[0] = 0
costs = [list(map(int, sys.stdin.readline().split())) for _ in range(n)]
for cost, guest in costs:
    for i in range(guest, len(dp)):
        dp[i] = min(dp[i-guest]+ cost, dp[i])

# c 이상의 고객을 늘리는 비용 중 최솟값
print(min(dp[c:]))

File: keemsebin/DynamicProgramming/BOJ_1106.js
const input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");

const [C, N] = input[0].split(" ").map(Number);
let arr = [];
for (let c = 0; c < N; c++) {
  arr.push(input[c + 1].split(" ").map(Number));
}

const dp = new Array(C + 100).fill(Infinity);
dp[0] = 0;

for (let i = 0; i < N; i++) {
  const [cost, customer] = arr[i];
  for (let j = customer; j < C + 100; j++) {
    dp[j] = Math.min(dp[j], dp[j - customer] + cost);
  }
}

console.log(Math.min(...dp.slice(C)));
